# Assignment 2: Linear Models and Validation Metrics (30 marks total)
### Due: October 10 at 11:59pm

### Name: Carissa Chung

### In this assignment, you will need to write code that uses linear models to perform classification and regression tasks. You will also be asked to describe the process by which you came up with the code. More details can be found below. Please cite any websites or AI tools that you used to help you with this assignment.

## Part 1: Classification (14.5 marks total)

You have been asked to develop code that can help the user determine if the email they have received is spam or not. Following the machine learning workflow described in class, write the relevant code in each of the steps below:

### Step 0: Import Libraries

In [26]:
import numpy as np
import pandas as pd

### Step 1: Data Input (1 mark)

The data used for this task can be downloaded using the yellowbrick library:
https://www.scikit-yb.org/en/latest/api/datasets/spam.html

Use the yellowbrick function `load_spam()` to load the spam dataset into the feature matrix `X` and target vector `y`.

Print the size and type of `X` and `y`.

In [27]:
# TO DO: Import spam dataset from yellowbrick library
from yellowbrick.datasets import load_spam
X, y = load_spam()

# TO DO: Print size and type of X and y
print("The size of X is:", X.size)
X.info()

print("\nThe size of y is:", y.size)
y.info()


The size of X is: 262200
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4600 entries, 0 to 4599
Data columns (total 57 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   word_freq_make              4600 non-null   float64
 1   word_freq_address           4600 non-null   float64
 2   word_freq_all               4600 non-null   float64
 3   word_freq_3d                4600 non-null   float64
 4   word_freq_our               4600 non-null   float64
 5   word_freq_over              4600 non-null   float64
 6   word_freq_remove            4600 non-null   float64
 7   word_freq_internet          4600 non-null   float64
 8   word_freq_order             4600 non-null   float64
 9   word_freq_mail              4600 non-null   float64
 10  word_freq_receive           4600 non-null   float64
 11  word_freq_will              4600 non-null   float64
 12  word_freq_people            4600 non-null   float64
 13  word_fre

### Step 2: Data Processing (1.5 marks)

Check to see if there are any missing values in the dataset. If necessary, select an appropriate method to fill-in the missing values.

In [28]:
# TO DO: Check if there are any missing values and fill them in if necessary
print(X.isna().sum())
print("\n", y.isna().sum())

word_freq_make                0
word_freq_address             0
word_freq_all                 0
word_freq_3d                  0
word_freq_our                 0
word_freq_over                0
word_freq_remove              0
word_freq_internet            0
word_freq_order               0
word_freq_mail                0
word_freq_receive             0
word_freq_will                0
word_freq_people              0
word_freq_report              0
word_freq_addresses           0
word_freq_free                0
word_freq_business            0
word_freq_email               0
word_freq_you                 0
word_freq_credit              0
word_freq_your                0
word_freq_font                0
word_freq_000                 0
word_freq_money               0
word_freq_hp                  0
word_freq_hpl                 0
word_freq_george              0
word_freq_650                 0
word_freq_lab                 0
word_freq_labs                0
word_freq_telnet              0
word_fre

For this task, we want to test if the linear model would still work if we used less data. Use the `train_test_split` function from sklearn to create a new feature matrix named `X_small` and a new target vector named `y_small` that contain **5%** of the data.

In [29]:
# TO DO: Create X_small and y_small
from sklearn.model_selection import train_test_split

# make separate subset of data
X_large, X_small, y_large, y_small = train_test_split(X, y, test_size=0.05, random_state=0, stratify=y)
# try with stratify=y

### Step 3: Implement Machine Learning Model

1. Import `LogisticRegression` from sklearn
2. Instantiate model `LogisticRegression(max_iter=2000)`.
3. Implement the machine learning model with three different datasets:
    - `X` and `y`
    - Only first two columns of `X` and `y`
    - `X_small` and `y_small`

In [47]:
from sklearn.linear_model import LogisticRegression

In [48]:
X_train, X_val, y_train, y_val = train_test_split(X, y, random_state=0)
linear_model_001 = LogisticRegression(max_iter=2000).fit(X_train, y_train)

In [49]:
X_2col = X.iloc[:, 0:2]
X_2col_train, X_2col_val, y2_train, y2_val = train_test_split(X_2col, y, random_state=0)
linear_model_002 = LogisticRegression(max_iter=2000).fit(X_2col_train, y2_train)

In [50]:
X_small_train, X_small_val, y_small_train, y_small_val = train_test_split(X_small, y_small, random_state=0)
linear_model_003 = LogisticRegression(max_iter=2000).fit(X_small_train, y_small_train)

### Step 4: Validate Model

Calculate the training and validation accuracy for the three different tests implemented in Step 3

### Step 5: Visualize Results (4 marks)

1. Create a pandas DataFrame `results` with columns: Data size, training accuracy, validation accuracy
2. Add the data size, training and validation accuracy for each dataset to the `results` DataFrame
3. Print `results`

In [34]:
# TO DO: ADD YOUR CODE HERE FOR STEPS 3-5
# Note: for any random state parameters, you can use random_state = 0
# HINT: USING A LOOP TO STORE THE DATA IN YOUR RESULTS DATAFRAME WILL BE MORE EFFICIENT

In [63]:
Training = []
Validation = []

# Test 1: X and y
Training.append(linear_model_001.score(X_train, y_train))
Validation.append(linear_model_001.score(X_val, y_val))

print("Test 1: X and y")
print("Training score: ", Training[0])
print("Validation score: ", Validation[0])

# Test 2: Only first two columns of X and y
Training.append(linear_model_002.score(X_2col_train, y2_train))
Validation.append(linear_model_002.score(X_2col_val, y2_val))

print("Test 2: First 2 cols of X and y")
print("Training score: ", Training[1])
print("Validation score: ", Validation[1])

# Test 3: X_small and y_small
Training.append(linear_model_003.score(X_small_train, y_small_train))
Validation.append(linear_model_003.score(X_small_val, y_small_val))

print("Test 3: X_small and y_small")
print("Training score: ", Training[2])
print("Validation score: ", Validation[2])



Test 1: X and y
Training score:  0.928695652173913
Validation score:  0.9373913043478261
Test 2: First 2 cols of X and y
Training score:  0.6084057971014493
Validation score:  0.6130434782608696
Test 3: X_small and y_small
Training score:  0.936046511627907
Validation score:  0.8620689655172413


In [68]:
columns = ["Test #", "Data size", "Training accuracy", "Validation accuracy"]
results = pd.DataFrame(columns=columns)
size = [X.size, X_2col.size, X_small.size]

for i in range(3):
    data = {
        columns[0]: i,
        columns[1]: size[i],
        columns[2]: Training[i],
        columns[3]: Validation[i]
    }

    results = pd.concat([results, pd.DataFrame([data])], ignore_index=True)

print(results)

  Test # Data size  Training accuracy  Validation accuracy
0      0    262200           0.928696             0.937391
1      1      9200           0.608406             0.613043
2      2     13110           0.936047             0.862069


### Questions (4 marks)
1. How do the training and validation accuracy change depending on the amount of data used? Explain with values.
2. In this case, what do a false positive and a false negative represent? Which one is worse?

1. Increasing data size appears to have a more significant effect on the validation accuracy. As shown for Test 1 using X and y, the validation accuracy was the highest at 94%. However, for Test 3 with a data size in between both Test 1 and Test 2, the validation accuraacy decreased to 86%. This observation can be further supported by a drop in the validation accuracy, 61%, for Test 2 with the smallest data size. Initially, increasing the data size may improve the validation accuracy as it's being evaluated on a larger and more representative dataset. At a certain dataset size, the validation accuracy will plateau and decrease which occurs because the validation set becomes more representative. This makes it harder for the model to overfit, but it also becomes harder for the model to fit the training data perfectly. This correlation however is not as clear for the training accuracy. The training accuracy for Test 1 with the largest data size was found to be 93%. However, when the data size decreased in Test 3, the training accuracy increased to 94%. However, it appears that a small data size will cause a decrease in both the training and validation accuracies. Despite that discrepancy, in most cases, as the amount of data increases, the training accuracy generally tends to increase. This is because the model has more examples to learn from which leads to better fitting of the underlying patterns in the data. At a certain point, increasing the data size will not significantly improve the training accuracy (may plateau) which can be seen in this case.

2. False positive represents email that have been classified as spam when they are actually not spam. False negative represents email that have not been classified as spam but are actually spam. In this case, false positives are worse. This is because some important emails may end up being deleted/hidden/ignored by the user since they have been marked as spam. For the case of a false negative, the user may become annoyed by having their inbox cluttered with spam emails but atleast no important information will be lost by an email being classified as spam when it actually is not.

### Process Description (4 marks)
Please describe the process you used to create your code. Cite any websites or generative AI tools used. You can use the following questions as guidance:
1. Where did you source your code?
1. In what order did you complete the steps?
1. If you used generative AI, what prompts did you use? Did you need to modify the code at all? Why or why not?
1. Did you have any challenges? If yes, what were they? If not, what helped you to be successful?

*DESCRIBE YOUR PROCESS HERE*
1. I referred to the class/lab notes to determine the functions required to fit the model and predict data. I also referenced the pandas pydata website to find out how to display the type of X and y (https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.info.html). I used ChatGPT (https://chat.openai.com) to check if my loop to create the results table was correct as I had been getting an error trying to run it. Turns out, the append method had been removed.
2. To create my code, I started from the first step and completed them in sequential order. Initially I had hardcoded the results table but added in a loop for a more efficient approach.
3. I had only used generative AI to help with checking my loop for creating the results table. I had initially used append but that method had been removed for pandas. Thus, ChatGPT was able to provide me with the updated method, concat, to add new rows of data to the results table based on the following prompt, "pandas.append has been removed. what is an alternative way to add a row to pandas dataframe?". I did not use the code provided by ChatGPT. Instead, I used the method ChatGPT provided and integrated that into my code. 
4. My biggest challenge was figuring out how to create the results table using a loop. I had initially wanted to prevent creating lists for the training and validation scores but in the end, it was the method that worked the best for me. In addition, another major challenge was trying to solve the environment issues. I was unable to read in the data and thus could not begin the assignment. Using Google Collab worked but I was determined to use VSCode. Thus, I recreated a new environment and updated all package versions. This solved the issue and I was able to complete the assignment via VSCode. 
I believe that reviewing the class notes and lab material helped to ensure that there were no major challenges in completing this assignment.

## Part 2: Regression (10.5 marks total)

For this section, we will be evaluating concrete compressive strength of different concrete samples, based on age and ingredients. You will need to repeat the steps 1-4 from Part 1 for this analysis.

### Step 1: Data Input (1 mark)

The data used for this task can be downloaded using the yellowbrick library:
https://www.scikit-yb.org/en/latest/api/datasets/concrete.html

Use the yellowbrick function `load_concrete()` to load the spam dataset into the feature matrix `X` and target vector `y`.

Print the size and type of `X` and `y`.

In [69]:
# TO DO: Import spam dataset from yellowbrick library
from yellowbrick.datasets import load_concrete
X, y = load_concrete()

# TO DO: Print size and type of X and y
print("The size of X is:", X.size)
X.info()
# print("The type of X is:", X.info())

print("\nThe size and type of y is:", y.size)
y.info()
# print("The type of y is:", y.info())

The size of X is: 8240
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1030 entries, 0 to 1029
Data columns (total 8 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   cement  1030 non-null   float64
 1   slag    1030 non-null   float64
 2   ash     1030 non-null   float64
 3   water   1030 non-null   float64
 4   splast  1030 non-null   float64
 5   coarse  1030 non-null   float64
 6   fine    1030 non-null   float64
 7   age     1030 non-null   int64  
dtypes: float64(7), int64(1)
memory usage: 64.5 KB

The size and type of y is: 1030
<class 'pandas.core.series.Series'>
RangeIndex: 1030 entries, 0 to 1029
Series name: strength
Non-Null Count  Dtype  
--------------  -----  
1030 non-null   float64
dtypes: float64(1)
memory usage: 8.2 KB


### Step 2: Data Processing (0.5 marks)

Check to see if there are any missing values in the dataset. If necessary, select an appropriate method to fill-in the missing values.

In [70]:
# TO DO: Check if there are any missing values and fill them in if necessary
print("Missing values in X:")
print(X.isna().sum())
print("\nMissing values in y:", y.isna().sum())

Missing values in X:
cement    0
slag      0
ash       0
water     0
splast    0
coarse    0
fine      0
age       0
dtype: int64

Missing values in y: 0


### Step 3: Implement Machine Learning Model (1 mark)

1. Import `LinearRegression` from sklearn
2. Instantiate model `LogisticRegression(max_iter=2000)`.
3. Implement the machine learning model with `X` and `y`

In [71]:
# TO DO: ADD YOUR CODE HERE
# Note: for any random state parameters, you can use random_state = 0

from sklearn.linear_model import LinearRegression

X_train, X_val, y_train, y_val = train_test_split(X, y, random_state=0)
lr = LinearRegression().fit(X_train, y_train)

### Step 4: Validate Model (1 mark)

Calculate the training and validation accuracy using mean squared error and R2 score.

In [72]:
# TO DO: ADD YOUR CODE HERE
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score

y_pred_train = lr.predict(X_train)
y_pred_val = lr.predict(X_val)

mse_train = mean_squared_error(y_train, y_pred_train)
mse_val = mean_squared_error(y_val, y_pred_val)

r2_train = r2_score(y_train, y_pred_train)
r2_val = r2_score(y_val, y_pred_val)

print("Training accuracy using MSE and R2:", mse_train, ",", r2_train)
print("Validation accuracy using MSE and R2:", mse_val, ",", r2_val)

Training accuracy using MSE and R2: 111.35843861132467 , 0.6108229424520555
Validation accuracy using MSE and R2: 95.90413603680645 , 0.6234144623633329


### Step 5: Visualize Results (1 mark)
1. Create a pandas DataFrame `results` with columns: Training accuracy and Validation accuracy, and index: MSE and R2 score
2. Add the accuracy results to the `results` DataFrame
3. Print `results`

In [73]:
# TO DO: ADD YOUR CODE HERE
results = pd.DataFrame()
results['Index'] = ['MSE', 'R2']
results['Training accuracy'] = [mse_train, r2_train]
results['Validation accuracy'] = [mse_val, r2_val]
results

Index  Training accuracy  Validation accuracy
0   MSE         111.358439            95.904136
1    R2           0.610823             0.623414

### Questions (2 marks)
1. Did using a linear model produce good results for this dataset? Why or why not?

I don't believe the linear model produced great results for this dataset. Looking at the R2 score, we would want that value to be as close to 1.0 as possible. However, for this dataset, the highest R2 value was only 0.62. Having a R2 score near 1.0 would indicate that the model is able to accurately predict the outcome of the dependable variables/features. Typically, if the R2 value for the training set is higher than the R2 score of the validation set, the model is overfitted. Since the R2 score for the validation set is slightly higher than the training, it could indicate that the model has not been overfitted and with a larger dataset, the validation R2 score could increase furthher.

For the mean squared error value, a larger MSE indicates that the predicted data are dispersed widely from the expected results, whereas a smaller MSE suggests the opposite. A smaller MSE is preferred because it indicates that your data points are were predicted to be more similar to the known outcome. The MSE is the average of all of the squared differences between the true values and the predicted values. Our goal is to select the model where the test MSE is lowest across choices of other models.

In addition, the dataset states that the concrete compressive strength (target) is a highly nonlinear function of age and ingredients. Based on this, it is reasonable to assume that a linear model would not be the best model to train to predict the target. 

### Process Description (4 marks)
Please describe the process you used to create your code. Cite any websites or generative AI tools used. You can use the following questions as guidance:
1. Where did you source your code?
1. In what order did you complete the steps?
1. If you used generative AI, what prompts did you use? Did you need to modify the code at all? Why or why not?
1. Did you have any challenges? If yes, what were they? If not, what helped you to be successful?

*DESCRIBE YOUR PROCESS HERE*
1. To determine the mean squared error and r2 score, I had to refer to the scikit-learn.org website to find the appropriate methods (https://scikit-learn.org/stable/modules/generated/sklearn.metrics.mean_squared_error.html and https://scikit-learn.org/stable/modules/generated/sklearn.metrics.r2_score.html). Similar to part 1, I had referred to the class and lab notes to reference which methods are required to fit and test the model.
2. Similar to part 1, I had completed the steps in sequential order. Most of the steps required the previous step to be completed or else it would be difficult to complete the next steps. 
3. No generative AI was used for this part of the assignment.
4. The biggest challenge for this part of the assisngment was interpreting the results. I had to refer to the class notes to determine the significance of the MSE and R2 score values. In addition, no loop was used for the results table for this part of the assignment since the results table is so small. Similar to the previous part, reviewing the course material was integral to ensuring no major challenges arose. 

## Part 3: Observations/Interpretation (3 marks)

Describe any pattern you see in the results. Relate your findings to what we discussed during lectures. Include data to justify your findings.


*ADD YOUR FINDINGS HERE*

Since MSE is a measure of the average squared difference between the actual (observed) values and the predicted values generated by the linear regression model, a lower MSE would indicate a better fit of the model to the data. MSE serves as a way to quantify how well the model's predictions match the actual data. In other words, smaller MSE values imply that the model's predictions are closer to the actual values, while larger MSE values indicate a poorer fit. 

On the other hand, the R2 score represents the variance proportion in the target which can be explained by the features used in the linear regression model (e.g., the goodness of the fit of the linear regression model). The R2 score ranges from 0 to 1, with 1 being an ideal model fitted to the dataset. An R2 score of 0 indicates that the model doesn't explain any variance, while an R2 score of 1 indicates a perfect fit where the model perfectly predicts the target variable.

As observed in the results from the model, a higher R2 score was observed for a lower MSE score (e.g., MSE of 111.36 had an associated R2 score of 0.61 whereas an MSE of 95.9 had an associated R2 score of 0.62.) This is expected as a lower MSE indicates that the predicted values deviated less from the target values. With regards to the R2 score, a higher score indicates a better fitted model to the data. 

## Part 4: Reflection (2 marks)
Include a sentence or two about:
- what you liked or disliked,
- found interesting, confusing, challangeing, motivating
while working on this assignment.


*ADD YOUR THOUGHTS HERE*
- I enjoyed having the opportunity to working through examples from start to finish. It was usefult to reinforce the material learned in class and during the lab sessions.
- I found the results to be interesting but confusing at the same time. Getting results that weren't quite as expected was helpful in getting me to try to understand why these particular results were observed. 

## Part 5: Bonus Question (4 marks)

Repeat Part 2 with Ridge and Lasso regression to see if you can improve the accuracy results. Which method and what value of alpha gave you the best R^2 score? Is this score "good enough"? Explain why or why not.

**Remember**: Only test values of alpha from 0.001 to 100 along the logorithmic scale.

In [94]:
# TO DO: ADD YOUR CODE HERE

# RIDGE REGRESSION
from sklearn.linear_model import Ridge
from sklearn.preprocessing import StandardScaler

X, y = load_concrete()
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

ridge_model_1 = Ridge(alpha=0.001)
ridge_model_1.fit(X_train, y_train)

y_pred_1 = ridge_model_1.predict(X_test)

mse_1 = mean_squared_error(y_test, y_pred_1)
r2_1 = r2_score(y_test, y_pred_1)

print("Ridge Regression with alpha = 0.001")
print(f"Mean Squared Error: {mse_1}")
print(f"R-squared (R2): {r2_1}")

ridge_model_2 = Ridge(alpha=1.0)
ridge_model_2.fit(X_train, y_train)

y_pred_2 = ridge_model_2.predict(X_test)

mse_2 = mean_squared_error(y_test, y_pred_2)
r2_2 = r2_score(y_test, y_pred_2)

print("\nRidge Regression with alpha = 1.0")
print(f"Mean Squared Error: {mse_2}")
print(f"R-squared (R2): {r2_2}")

ridge_model_3 = Ridge(alpha=100.0)
ridge_model_3.fit(X_train, y_train)

y_pred_3 = ridge_model_3.predict(X_test)

mse_3 = mean_squared_error(y_test, y_pred_3)
r2_3 = r2_score(y_test, y_pred_3)

print("\nRidge Regression with alpha = 100.0")
print(f"Mean Squared Error: {mse_3}")
print(f"R-squared (R2): {r2_3}")

ridge_model_4 = Ridge(alpha=10.0)
ridge_model_4.fit(X_train, y_train)

y_pred_4 = ridge_model_4.predict(X_test)

mse_4 = mean_squared_error(y_test, y_pred_4)
r2_4 = r2_score(y_test, y_pred_4)

print("\nRidge Regression with alpha = 10.0")
print(f"Mean Squared Error: {mse_4}")
print(f"R-squared (R2): {r2_4}")


# LASSO REGRESSION
from sklearn.linear_model import Lasso

lasso_model_1 = Lasso(alpha=0.001)
lasso_model_1.fit(X_train, y_train)
y_pred_l1 = lasso_model_1.predict(X_test)

mse_l1 = mean_squared_error(y_test, y_pred_l1)
r2_l1 = r2_score(y_test, y_pred_l1)

print("\nLasso Regression with alpha = 0.001")
print(f"Mean Squared Error: {mse_l1}")
print(f"R-squared (R2): {r2_l1}")

lasso_model_2 = Lasso(alpha=0.1)
lasso_model_2.fit(X_train, y_train)
y_pred_l2 = lasso_model_2.predict(X_test)

mse_l2 = mean_squared_error(y_test, y_pred_l2)
r2_l2 = r2_score(y_test, y_pred_l2)

print("\nLasso Regression with alpha = 0.1")
print(f"Mean Squared Error: {mse_l2}")
print(f"R-squared (R2): {r2_l2}")


Ridge Regression with alpha = 0.001
Mean Squared Error: 95.97547158704613
R-squared (R2): 0.627541655086022

Ridge Regression with alpha = 1.0
Mean Squared Error: 95.96966712154402
R-squared (R2): 0.6275641808582665

Ridge Regression with alpha = 100.0
Mean Squared Error: 101.81240449751492
R-squared (R2): 0.6048898844277723

Ridge Regression with alpha = 10.0
Mean Squared Error: 96.27676054442806
R-squared (R2): 0.6263724231504898

Lasso Regression with alpha = 0.001
Mean Squared Error: 95.9746068931565
R-squared (R2): 0.6275450107606507

Lasso Regression with alpha = 0.1
Mean Squared Error: 96.41516117745807
R-squared (R2): 0.6258353226824118


*ANSWER HERE*

Both methods gave very similar results. At higher values of alpha, more coefficients are driven to zero (Lasso) or be small (Ridge) resulting in only retaining the most important features in the model. Thus, the model becomes simpler and less prone to overfitting. However, the model could underfit the data if the regularization is too strong. A smaller alpha value indicates less coefficients to be driven to zero or minimzed thus, resulting in a more complicated model. 

In both cases, the R2 score is not really acceptable as the value is quite far from the desired value near 1.0. This R2 value indicates that the model is only able to predict the correct values ~63% of the time.